### PR TITLE
Include unindexed field in FieldStats response

### PR DESCRIPTION
**Pull request by jimczi**
This change adds non-searchable fields to the FieldStats response. These fields do not have min/max informations but they can be aggregatable. Fields that are only stored in _source (store:no, index:no, doc_values:no) will still be missing since they do not have any useful information to show. Indices and clients must be at least on V_5_2_0 to see this change.

Closes https://github.com/elastic/elasticsearch/issues/21952

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: Add Javadoc for these two ctors now?

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: It is weird to see this pre-NamedWriteable stuff still around. It might make sense to convert these to NamedWriteable in master.

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: `writeMinMax` isn't going to work for this one, I think.

File: core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java]: I think this'll be more readable if you assign `result.getAllFieldStats().get(field)` to a variable.

File: core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java]: I'm not really a fan of `assertThat(..., equalTo(...))`. I feel like it is usually more clear to do `assertEquals(...)`. Like, if you have lots of assertions and you think they should line up then fine, but in this case I think they are all `equalTo` so maybe switch? Not a big deal either way.

File: core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java]: I'd test both in the same test run because it is fast.

File: core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java]: I know it isn't a part of this change but I'm not a fan of the `int type = randomInt(5); switch (type) {...}` idiom because I think it is easy to make a mistake that excludes a branch. I'd prefer `randomFrom(() -> {...}, () -> {...}).get()` because it is more obvious that you haven't forgot a branch. Is that crazy?

Comment (jimczi) [inline core/src/test/java/org/elasticsearch/fieldstats/FieldStatsTests.java]: I think it's fine to use the switch statement as long as the `default` throws an exception ?

**Review by nik9000 — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: I am not sure of this, NamedWriteable is ok when you want to serialize objects that have nothing in common but that's not the case here. We just want to serialize the min/max value differently for each specialization of the class. Rewriting this as a NamedWriteable would be completely different. We would need to create a class to encapsulate the min/max info and then use NamedWriteable to serialize these classes. I can do that but I prefer to do it in another PR since it's not related to the change that we're trying to make here ?

**Comment by jimczi** [comment]
Thanks @nik9000 
I pushed a commit to address your comments.
If it's ok for you I'd like to push this change (modulo the changes that you requested) and do the NamedWriteable stuff in another PR.

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: Yeah, another PR is totally fine. No need to hold this up. And I'm not 100% sure it makes sense to convert them, just that it makes sense to think about it.

We have lots of NamedWriteables that write things in common - queries and aggregations do it all the time. They defer to superclasses to load the common stuff.

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: it feels wrong to say that we know the min/max unless the information is available for all indices?

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: should we throw an exception if min/max info is not available?

**Review by jpountz — COMMENTED** [review]
I left a question.

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: I pushed https://github.com/elastic/elasticsearch/pull/21821/commits/be820240a27ce57e0fd825728b3dab11276b0a6d

Though I think we should distinguish null values because the shard is empty and null values because the field is not indexed. I could not do it in this commit because SeqNoFieldType#stats computes min/max values on a non-searchable field.
@bleskes @jasontedor what is the plan for this SeqNoFieldType#stats ? Are we going to index this field and remove the min/max retrieval ?

File: core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: This is the current behavior for fields that are not indexed so maybe better to just ignore these fields for now ? The index constraint works on indexed fields and always returns false on non-indexed fields, I can add a sentence like this in the docs if needed ?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java]: That works for me. Thanks for the explanation.

**Review by jpountz — APPROVED** [review]
LGTM

**Comment by jimczi** [comment]
Thanks @jpountz and @nik9000 !
I'll merge this to 5.2.0 now

**Pull request merged**
